### PR TITLE
[Dynamic Buffer Calc] Enhance dynamic buffer calculation test

### PR DESCRIPTION
**Pull request by stephenxs**
### Description of PR

Summary:
Enhance the dynamic buffer calculation test

Signed-off-by: Stephen Sun <stephens@nvidia.com>
### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

#### What is the motivation for this PR?
Enhance the dynamic buffer calculation test:
- Ensure the buffer pool size is recovered after all the configurations have been reverted
  - The buffer pool size is verified based on the value fetched from the switch at the beginning of the test. Afterwards, it will calculate the new buffer pool size according to the configuration change.
  - The configurations changed during the test will be restored at the end of the test and the switch will recalculate the buffer pool size according to the restored value. However, it can take a few seconds for buffer pool size to be recovered in rare conditions.
  - If the next test starts immediately after the previous one finished, it is possible that the buffer pool size is not fully recovered, which results in the next test verifies the buffer pool size based on a wrong value, which fails the test.
- Adjust the buffer pool size calculation while toggling the shared headroom pool and there are 400G ports on the system.
- Add test case for removing PGs when a port is administratively down.
- Bug fixes:
  - in `check_buffer_profiles_for_shp`: the check passes only if all lossless profiles satisfy the condition
  - in QoS test, the shared headroom pool size should be fetched from APPL_DB or CONFIG_DB according to whether buffer information is in APPL_DB or CONFIG_DB

#### How did you do it?
- Ensure the buffer pool size is recovered after all the configurations have been reverted
  - Change the scope of the methods `_check_pool_size` and `_get_pool_size_from_asic_db` from `check_pool_size` to global so that they can be called by other methods.
  - Check the buffer pool size after all configuration changes have been restored at the end of each test.
- Adjust the buffer pool size calculation while toggling the shared headroom pool and there are 400G ports on the system.
  - Introduce a new method `_fetch_size_difference_for_400g_ports` to calculate the difference in buffer pool sizes between the modes with/without shared headroom pool enabled.
  - Take the difference into consideration when calculating the buffer pool size during toggling the shared headroom pool
- Add test case for removing PGs when a port is administratively down.
  - Introduce a new method `test_port_admin_down` for it.

#### How did you verify/test it?
Run the regression test.

### CODE REVIEW

**Comment by stephenxs** [comment]
> This pull request **introduces 1 alert** when merging [4de0a61](https://github.com/Azure/sonic-mgmt/commit/4de0a615f1e2af1154ee645ae189df023d754fff) into [504fd87](https://github.com/Azure/sonic-mgmt/commit/504fd87354e8abd56a7fa5ff669b94d2f405b107) - [view on LGTM.com](https://lgtm.com/projects/g/Azure/sonic-mgmt/rev/pr-43c8535c8a9decdcb53f394c1ebac7dfbe856cab)
> 
> **new alerts:**
> 
> * 1 for Variable defined multiple times

Fixed.

**Comment by stephenxs** [comment]
> This pull request **introduces 1 alert** when merging [6f17fef](https://github.com/Azure/sonic-mgmt/commit/6f17fefc45cc56b9933baf5fd519dfd242f2034c) into [78d2df9](https://github.com/Azure/sonic-mgmt/commit/78d2df9c82f6cb4964126af634c0bda2238e5f8c) - [view on LGTM.com](https://lgtm.com/projects/g/Azure/sonic-mgmt/rev/pr-f2cecd9e3f2b4c484ee8cc2f9672cba6e49fdcab)
> 
> **new alerts:**
> 
> * 1 for Unused local variable

Fixed

**Comment by stephenxs** [comment]
@neethajohn could you please review it? Thanks

File: tests/qos/test_buffer.py [hunk 1]
```diff
@@ -200,6 +234,21 @@ def check_pool_size(duthost, ingress_lossless_pool_oid, **kwargs):
...
+        Args:
+            duthost: The duthost object
+            conn_graph_facts: The connection graph facts object
+        """
+        hostname = conn_graph_facts['device_conn'].keys()[0]
+        ports_info = conn_graph_facts['device_conn'][hostname]
+        ports_400g = [port for port in ports_info.keys() if ports_info[port]['speed'] == '400000']
+
+        lossless_pgs = duthost.shell('redis-cli keys "BUFFER_PG_TABLE:Ethernet*:3-4"')['stdout'].split()
+        lossless_pgs_400g = [pg for pg in lossless_pgs if pg[16:-4] in ports_400g]
```
[the remark this answers is not in the document]
Comment (stephenxs) [reply tests/qos/test_buffer.py:249]: A lossless pg is like this `BUFFER_PG_TABLE:Ethernet8:3-4`. The prefix and suffix have fixed lengths but the port name has a variable length, so pg[16:-4] is the port name.

**Review by neethajohn — APPROVED** [review]

**Pull request merged**